src/tms_service.ts

```diff
@@ -0,0 +1,296 @@
+};
+
+type EmsSprite = {
+  height: number;
+  pixelRatio: number;
+  width: number;
+  x: number;
+  y: number;
+};
+
+type EmsSpriteSheet = {
+  [spriteName: string]: EmsSprite;
+};
+
+type EmsRasterStyle = {
+  tilejson: string;
+  name: string;
+  attribution: string;
+  minzoom: number;
+  maxzoom: number;
+  bounds: number[];
+  format: string;
+  type: string;
+  tiles: string[];
+  center: number[];
+};
+
+export class TMSService extends AbstractEmsService {
+  protected readonly _config: TMSServiceConfig;
+
+  _getRasterStyleJson = _.once(
+    async (): Promise<EmsRasterStyle | undefined> => {
+      const rasterUrl = this._getStyleUrlForLocale('raster');
+      if (rasterUrl) {
+        const url = this._proxyPath + this._getAbsoluteUrl(rasterUrl);
+        return this._emsClient.getManifest<EmsRasterStyle>(
+          this._emsClient.extendUrlWithParams(url)
+        );
+      } else {
+        return;
+      }
+    }
+  );
+
+  _getVectorStyleJsonRaw = _.once(
+    async (): Promise<EmsVectorStyle | undefined> => {
+      const vectorUrl = this._getStyleUrlForLocale('vector');
+      if (vectorUrl) {
+        const url = this._proxyPath + this._getAbsoluteUrl(vectorUrl);
+        const vectorJson = await this._emsClient.getManifest<EmsVectorStyle>(
+          this._emsClient.extendUrlWithParams(url)
+        );
+        return { ...vectorJson };
+      } else {
+        return;
+      }
+    }
+  );
+
+  _getVectorStyleJsonInlined = _.once(
+    async (): Promise<EmsVectorStyle | undefined> => {
+      const vectorJson = await this._getVectorStyleJsonRaw();
+      if (vectorJson) {
+        const inlinedSources: EmsVectorSources = {};
+        const { sources } = vectorJson;
+        for (const sourceName of Object.getOwnPropertyNames(sources)) {
+          const { url } = sources[sourceName];
+          const sourceUrl = this._proxyPath + this._getAbsoluteUrl(url);
+          const extendedUrl = this._emsClient.extendUrlWithParams(sourceUrl);
+          const sourceJson = await this._emsClient.getManifest<EmsVectorSource>(extendedUrl);
+
+          const extendedTileUrls = sourceJson.tiles.map((tileUrl: string) => {
+            const url = this._proxyPath + this._getAbsoluteUrl(tileUrl);
+            return this._emsClient.extendUrlWithParams(url);
+          });
+          // Override the attribution in the sources with the localized attribution
+          const htmlAttribution = await this.getHTMLAttribution();
+          inlinedSources[sourceName] = {
+            ...sourceJson,
+            attribution: htmlAttribution,
+            tiles: extendedTileUrls,
+          };
+        }
+        return {
+          ...vectorJson,
+          sources: inlinedSources,
+          sprite: await this._getSpriteSheetRootPath(),
+          glyphs: await this._getUrlTemplateForGlyphs(),
+        };
+      } else {
+        return;
+      }
+    }
+  );
+
+  constructor(config: TMSServiceConfig, emsClient: EMSClient, proxyPath: string) {
```

Comment:
For readability, I'm used to having some order in the classes, like first fields, then constructor, public, protected, and private methods; That's the `field-first` spec on the [TSLint rule](https://palantir.github.io/tslint/rules/member-ordering/) for this topic.